buku

```diff
@@ -1412,14 +1419,12 @@ class BukuDb:
 
         Returns
         -------
-        list or None
-            List of search results, or None if no matches.
+        list
+            List of search results.
```

Comment:
Currently third-party callers check for None. Will that work with this?

Reply:
As per [my original argumentation](https://github.com/jarun/buku/issues/654):

* The claim "this function returns `None` when there's no matches" was not accurate in the first place: `None` was returned _only in certain cases_, outside of them the returned "no matches" result was `[]`.  
  (More specifically, the "certain cases" thing concerns specific checks in `searchdb()` and `search_by_tag()`; as for `get_all_rec()`, `search_keywords_and_filter_by_tags()` and `exclude_results_from_search()` – they _always_ [[1]](https://github.com/jarun/buku/commit/e5395b5528b85a2dbcde0d48882bfe280dfddf00#diff-9484ce875eae0f1affbabc2b6e38781a43629ec0a0fd083bfdfdfa8fbdb1d39a) [[2]](https://github.com/jarun/buku/commit/6db194405b293791cf532ee0301095fa7359173b#diff-9484ce875eae0f1affbabc2b6e38781a43629ec0a0fd083bfdfdfa8fbdb1d39a) [[3]](https://github.com/jarun/buku/commit/e12d2f95f067f44276413455424af78796809994#diff-9484ce875eae0f1affbabc2b6e38781a43629ec0a0fd083bfdfdfa8fbdb1d39a) returned a `list()`, the latter two explicitly and the former [by docs of the returned call](https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.fetchall).)
* Both `[]` and `None` are falsey, idiomatically for Python – checking specifically for `is None` makes little here when any conditional can simply accept the value itself without losing out on correctness (rather, doing a regular falsiness check is more correct, because `[]` is returned by all of these functions).
* Returning `[]` instead of `None` makes processing of such results simpler, when there's a need to do something more complex than an emptiness check (like how it's done in `search_keywords_and_filter_by_tags()`, which previously needed workarounds for the `None` value).

(…Somehow I keep getting the feeling that this code was written in C originally then translated into Python directly without accounting for the language differences. :sweat_smile:)

Reply:
Once again, if the caller checks for None as in the existing API, will this work or break? Does the caller need to change? That's my greatest concern.

Reply:
The thing is, _the **existing API** does not return values as described in docs_.

If the caller checks specifically for `is None` (as opposed to the idiomatic falsiness check that makes more sense when dealing with functions returning a list) in `searchdb()` or `search_by_tag()`, then he's _not_ checking for _empty search results_ (**which are returned as `[]` in the current version**) but _specifically_ for _invalid input passed to the function_. Since input validation isn't a part of the documented API, there's no problem in changing it out.

And if the caller checks for `is None` in `get_all_rec()`, `search_keywords_and_filter_by_tags()` or `exclude_results_from_search()`, his code will keep not working _just as it had from the very beginning_ (because these functions **never returned `None` in the first place**).